### PR TITLE
Support for Elasticsearch 2.x

### PR DESCRIPTION
**Pull request by akonczak**
Support for ES 2.x - Spring Data ES has dependency with this pull  - DATAES-219

### CODE REVIEW

**Comment by Shredder121** [comment]
Thanks for the PR.
One idea: use the `com.querydsl.elasticsearch2` package, to prevent clashes with ClassLoaders not liking two classes with the same name and package.

File: querydsl-elasticsearch2/pom.xml [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+      <scope>provided</scope>
+    </dependency>
+
+  </dependencies>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>com.springsource.bundlor</groupId>
+        <artifactId>com.springsource.bundlor.maven</artifactId>
```
Comment (Shredder121) [inline querydsl-elasticsearch2/pom.xml]: `<version />` not present.
I notice that it's not there on the original elasticsearch module. Thoughts @timowest ?

Comment (Shredder121) [inline querydsl-elasticsearch2/pom.xml]: Maybe put it in a parent pom?

File: querydsl-elasticsearch2/src/test/java/com/querydsl/elasticsearch/PackageVerification.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import com.querydsl.apt.QuerydslAnnotationProcessor;
+import com.querydsl.codegen.CodegenModule;
+import com.querydsl.core.types.Expression;
+
+public class PackageVerification {
+
+    @Test
+    public void Verify_Package() throws Exception {
+        String version = System.getProperty("version");
+        verify(new File("target/querydsl-elasticsearch-" + version + "-apt-one-jar.jar"));
```
Comment (Shredder121) [inline querydsl-elasticsearch2/src/test/java/com/querydsl/elasticsearch/PackageVerification.java]: `target/querydsl-elasticsearch2-" + version + "-apt-one-jar.jar`

File: querydsl-elasticsearch2/pom.xml [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            <phase>verify</phase>
+            <configuration>
+              <systemProperties>
+                <property>
+                  <name>version</name>
+                  <value>${project.version}</value>
+                </property>
+              </systemProperties>
+              <includes>
+                <include>com/mysema/query/PackageVerification.java</include>
```
Comment (Shredder121) [inline querydsl-elasticsearch2/pom.xml]: `com/querydsl/elasticsearch/PackageVerification.java`

But I recommend having this under `com.querydsl.elasticsearch2/PackageVerification.java`

@timowest I believe this is also wrong in the original `querydsl-elasticsearch` project.
Its `javax.annotation.processing.Processor` file is still outdated.

I'll open up a PR.

**Comment by akonczak** [comment]
Additional changes: 
-renamed package to not clash with ES 1.x
-added version for bundler (main pom)
-changed paths for PackageVerification in all poms to reflect correct location

**Comment by Shredder121** [comment]
I forgot to change some `.` to `/` in my comment, so that should be `com/querydsl/elasticsearch2/PackageVerification.java`
But I first want to hear @timowest's opinion on some matters.

Anyway, thank you for spearheading this.

**Comment by kevinleturc** [comment]
Hi guys, thanks to review this PR on both projects ;).
I'll investigate a bit on spring bundlor plugin, I'm not sure if it's needed.

**Comment by kevinleturc** [comment]
Spring bundlor plugin seems to be not used during a classic compilation.
It is also present in dynamodb and hazelcast. Maybe we can remove it ?

**Comment by akonczak** [comment]
@Shredder121 

My mistake.
PackageVerification is only available only for ES/ES2, moreover test(s) will fails if you run it.

`Running com.querydsl.elasticsearch.PackageVerification
Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 0.118 sec <<< FAILURE! - in com.querydsl.elasticsearch.PackageVerification
Verify_Package(com.querydsl.elasticsearch.PackageVerification)  Time elapsed: 0.116 sec  <<< FAILURE!
org.junit.ComparisonFailure: expected:<com.[querydsl].apt.QuerydslAnnotat...> but was:<com.[mysema.query].apt.QuerydslAnnotat...>
	at org.junit.Assert.assertEquals(Assert.java:115)
	at org.junit.Assert.assertEquals(Assert.java:144)
	at com.querydsl.elasticsearch.PackageVerification.verify(PackageVerification.java:51)
	at com.querydsl.elasticsearch.PackageVerification.Verify_Package(PackageVerification.java:37)
`

**Comment by Shredder121** [comment]
Yes, `PackageVerification` is a sanity check, a safety net if you will.
It was supposed to run all the time, but it didn't..

File: querydsl-elasticsearch2/src/apt/META-INF/services/javax.annotation.processing.Processor [hunk 4]
```diff
@@ -0,0 +1 @@
+com.mysema.query.apt.QuerydslAnnotationProcessor
```
Comment (Shredder121) [inline querydsl-elasticsearch2/src/apt/META-INF/services/javax.annotation.processing.Processor]: This should point to `com.querydsl.apt.QuerydslAnnotationProcessor`.

**Comment by akonczak** [comment]
added PackageVerification for all module + basic test to verify that core classes are available.

**Comment by Shredder121** [comment]
@kevinleturc That particular plugin can indeed be removed.

In the `querydsl-root` pom we have configured the `org.apache.felix:maven-bundle-plugin`, so that should actually be used there.
```xml
<plugin>
  <groupId>org.apache.felix</groupId>
  <artifactId>maven-bundle-plugin</artifactId>
</plugin>
```

**Comment by akonczak** [comment]
i will do that later today

**Comment by Shredder121** [comment]
That's okay.

**Comment by akonczak** [comment]
ok all done.

**Comment by akonczak** [comment]
Do we have any date when this pull request can be approved ?

**Pull request merged**